keras/backend/tensorflow_backend.py

```diff
@@ -2580,6 +2583,10 @@ def _call(self, inputs):
                                 symbol_vals,
                                 session)
         fetched = self._callable_fn(*array_vals)
+        if self._merged_summaries is not None:
+            # self._writer and self._merged_summaries and self._epoch
+            # are assigned by TensorBaord callback
+            self._writer.add_summary(fetched[-1], self._epoch)
```

Comment:
`_merged_summaries` is added to fetches when the callable is created for the first time. That means the summary are computed at every call, even when `_merged_summaries` is None (but only added to the writer when  `_merged_summaries` is not None).

Also it means that if `_merged_summaries` is None when the callable is created, `_merged_summaries` will always be ignored henceforth.

You need to store `_merged_summaries` as a Function attribute and refresh the callable when it has changed (like we do for input tensors now).

Reply:
BTW, if this issue isn't being detected in tests then I suspect there isn't sufficient test coverage.

Reply:
I see. That's a really big problem. :(

I have a suggestion in contrast with keeping `_merged_summaries`. I'm not sure if it's good or not because I'm not aware of the extension of `fetches`, but here it goes:
We keep track of changes in `fetches` (using `_fetches`, pretty much like `_feed_symbols`) and its results. We could then simply add the summaries to it and implement the TensorBoard callback somewhat like this:

```python
def on_epoch_start(...):
    if freq and epoch % freq == 0:
        self.test_function.fetches += [self.merged]

def on_batch_end(...):
    if freq and epoch % freq == 0:
        summary = self.test_function.fetches_outputs[self.merged]
        self.writer.add_summary(summary, epoch)

def on_epoch_end(...):
    if freq and epoch % freq == 0:
        self.writer.flush()
        self.test_function.fetches.remove(self.merged)
        del self.test_function.fetches_outputs[self.merged]

def on_train_end(...):
    self.writer.close()
```

Reply:
Oh, I just realized `test_loop` doesn't wrap its `test_function` calls with `on_batch_start/end`, so this wouldn't work like this. A new `on_valid_batch_end` event would fix this, but I guess that'd be adding more complications than fixing here...